src/main/main.c

```diff
@@ -1235,11 +1236,18 @@ void re_thread_leave(void)
 		DEBUG_WARNING("re_thread_leave: re not ready\n");
 		return;
 	}
-	/* Dummy async event, to ensure timers are properly handled */
-	if (re->async)
-		re_thread_async(NULL, NULL, NULL);
-	re_atomic_rlx_set(&re->thread_enter, false);
-	re_unlock(re);
+
+	/* leave if on re thread */
+	if (re_atomic_rlx(&re->thread_enter) ||
+	    thrd_equal(re->tid, thrd_current())) {
```

Comment:
I think the check must be:
```suggestion
	    !thrd_equal(re->tid, thrd_current())) {
```

Reply:
I just copied the test from `re_thread_check` function.  According to it, we are on re thread if re_atomic_rlx(&re->thread_enter) or thrd_equal(re->tid, thrd_current().  Is there a bug in that function?